### PR TITLE
Move UnresolvedAttribute and Function resolution to QL

### PR DESCRIPTION
**Pull request by bpintea**
This moves the resolution of the UnresolveAttribute logic to QL, as well as the function resolution, making them available in the other *QL packages.

### CODE REVIEW

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/analyzer/AnalyzerRules.java [hunk 1]
```diff
@@ -86,4 +104,139 @@ protected LogicalPlan rule(LogicalPlan plan) {
...
+
+        protected Function resolveFunction(UnresolvedFunction uf) {
+            if (uf.analyzed()) {
+                return uf;
+            }
+            if (uf.childrenResolved() == false) {
+                return uf;
+            }
+
+            Objects.requireNonNull(functionRegistry);
```
Comment (astefan) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/analyzer/AnalyzerRules.java:132]: Again, I don't think this is necessary.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java [hunk 2]
```diff
@@ -901,17 +818,15 @@ private Expression replaceAliases(Expression condition, List<? extends NamedExpr
...
+        ResolveFunctions(Configuration configuration, FunctionRegistry functionRegistry) {
+            super(configuration, functionRegistry);
+        }
 
         @Override
         protected LogicalPlan rule(LogicalPlan plan) {
             return plan.transformExpressionsUp(UnresolvedFunction.class, uf -> {
-                if (uf.analyzed()) {
-                    return uf;
-                }
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java:911]: I think this section needs to stay.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/analyzer/AnalyzerRules.java [hunk 3]
```diff
@@ -86,4 +104,139 @@ protected LogicalPlan rule(LogicalPlan plan) {
...
     }
+
+    public static class ResolveFunctions extends AnalyzerRule<LogicalPlan> {
+
+        private final Configuration configuration;
+        private final FunctionRegistry functionRegistry;
+
+        public ResolveFunctions(Configuration configuration, FunctionRegistry functionRegistry) {
+            this.configuration = configuration;
+            Objects.requireNonNull(functionRegistry);
```
Comment (astefan) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/analyzer/AnalyzerRules.java:115]: I don't think this is necessary. Both these objects are essential and part of the core classes of both EQL and SQL. If until this point they are null, the NPE would have happened already.

**Review by astefan — APPROVED** [review]
LGTM. Left two minor comments, though.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java [hunk 4]
```diff
@@ -901,17 +818,15 @@ private Expression replaceAliases(Expression condition, List<? extends NamedExpr
         }
     }
 
-    private class ResolveFunctions extends AnalyzerRule<LogicalPlan> {
+    private static class ResolveFunctions extends AnalyzerRules.ResolveFunctions {
+
+        ResolveFunctions(Configuration configuration, FunctionRegistry functionRegistry) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java:823]: The configuration is per request while the Rule across requests. This needs to be broken out so `ResolveFunctions` is abstract and each QL project extends it as part of their own Analyzer and thus have access to their configuration.
Moving forward we could parameterized this so the Analyzer (and its rules) become singleton.

**Review by costin — CHANGES_REQUESTED** [review]
The current ResolveFunction rule does not depend on the configuration  (or the function registry) - neither should the common class; making it abstract keeps the common code in one place and lets actual implementations to plug in their instances in.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java (hunk 4 above)
Comment (bpintea) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]: Reworked after the parameterised rule and executor changes.

**Review by costin — APPROVED** [review]
thanks for spending the time on this - LGTM!

**Pull request merged**